In [1]:
!pip install requests

In [13]:
import pandas as pd
import numpy as np
import requests
from pandas import option_context

# Анализ исходных данных и построение пайпа по обогащению для поиска

В рамках данных работы посмотрим на исходные данные и выработаем стратегию для сбора дополнительных фичей, позволяющих построить поиск, с учетом ограничений по железу команды

## Предварительный анализ

Первым делом рассмотрим, какие данные были получены на вход

In [16]:
dataset = pd.read_csv('../source_dataset.csv')

with option_context('display.max_colwidth', 400, "display.max_rows", 100):
    display(dataset.head(100))

link  \
0   https://cdn-st.rutubelist.ru/media/b0/e9/ef285e0241139fc611318ed33071/fhd.mp4   
1   https://cdn-st.rutubelist.ru/media/39/6c/b31bc6864bef9d8a96814f1822ca/fhd.mp4   
2   https://cdn-st.rutubelist.ru/media/e9/e0/b47a9df14a5e97942715e5e705c0/fhd.mp4   
3   https://cdn-st.rutubelist.ru/media/87/43/b11df3f344d0af773aac81e410ee/fhd.mp4   
4   https://cdn-st.rutubelist.ru/media/d1/e7/642dc2194fcdb69664f832d5f2dd/fhd.mp4   
5   https://cdn-st.rutubelist.ru/media/e2/97/f9164f8a41479f961d64842154a7/fhd.mp4   
6   https://cdn-st.rutubelist.ru/media/0f/48/8a1ff7324073947a31e80f71d001/fhd.mp4   
7   https://cdn-st.rutubelist.ru/media/19/9c/98de3b4a4d9e83cb5dda828f899e/fhd.mp4   
8   https://cdn-st.rutubelist.ru/media/bf/ee/85fb6b80491db79e8baebe5c9d80/fhd.mp4   
9   https://cdn-st.rutubelist.ru/media/fc/73/d86600d04e519178a55774dfe65e/fhd.mp4   
10  https://cdn-st.rutubelist.ru/media/17/ae/a36f6f824fa1a67496d299c8ac6e/fhd.mp4   
11  https://cdn-st.rutubelist.ru/media/4b/54/0a36f96d4030a0f6ad1b6d7f4902/fhd.mp4   
12  https://cdn-st.rutubelist.ru/media/8c/58/96608b7a4834883ea2dd1c884c63/fhd.mp4   
13  https://cdn-st.rutubelist.ru/media/2f/4f/6b969c8c4a2aafcfca057b2a99a2/fhd.mp4   
14  https://cdn-st.rutubelist.ru/media/9f/ca/d5d0dedf44418e1a2b7d5cebe0e8/fhd.mp4   
15  https://cdn-st.rutubelist.ru/media/a4/d2/8e9737fd445faf0a20151f8c39f5/fhd.mp4   
16  https://cdn-st.rutubelist.ru/media/a3/9f/2352de2748b3868df583d51a402b/fhd.mp4   
17  https://cdn-st.rutubelist.ru/media/ac/08/3604604b4b8b89e34db1e0e91863/fhd.mp4   
18  https://cdn-st.rutubelist.ru/media/5a/35/65c80c3047d08634922cb22f1fee/fhd.mp4   
19  https://cdn-st.rutubelist.ru/media/03/eb/47329b1943349255453b6cec625d/fhd.mp4   
20  https://cdn-st.rutubelist.ru/media/28/b0/88ac63a546cb8174020a95470591/fhd.mp4   
21  https://cdn-st.rutubelist.ru/media/bf/6a/f040f0dd4afc90b8eb12c8d76571/fhd.mp4   
22  https://cdn-st.rutubelist.ru/media/73/48/47a9e25b4b36bc6254d29f42787a/fhd.mp4   
23  https://cdn-st.rutubelist.ru/media/fe/86/ab9d0a504c2a892c88dc02b79b56/fhd.mp4   
24  https://cdn-st.rutubelist.ru/media/86/0f/4a6b06514445bbe41cb62a5f279d/fhd.mp4   
25  https://cdn-st.rutubelist.ru/media/1e/39/f6e45a2142b095d96b286744441a/fhd.mp4   
26  https://cdn-st.rutubelist.ru/media/0d/97/531d4356476a9deb64239aed758a/fhd.mp4   
27  https://cdn-st.rutubelist.ru/media/85/3c/4ade487d46cfb36205c43e4dca40/fhd.mp4   
28  https://cdn-st.rutubelist.ru/media/50/0a/a56eb2824b93a57c3f4ad7e02ddd/fhd.mp4   
29  https://cdn-st.rutubelist.ru/media/a7/ff/3d901b774fc8b5728833abe9485d/fhd.mp4   
30  https://cdn-st.rutubelist.ru/media/b3/e2/07e47c1e4af8a1ea0a932b5cdae3/fhd.mp4   
31  https://cdn-st.rutubelist.ru/media/39/8a/ce3d8db142e38d99c708f406cca0/fhd.mp4   
32  https://cdn-st.rutubelist.ru/media/c5/74/92d9fe2a4fd99509c72ebf46e7d7/fhd.mp4   
33  https://cdn-st.rutubelist.ru/media/95/65/d45056f64cd48e919324c2ade667/fhd.mp4   
34  https://cdn-st.rutubelist.ru/media/19/b5/aa6a2f7b46ad81e3c78e947d472f/fhd.mp4   
35  https://cdn-st.rutubelist.ru/media/03/b6/7aab45314f66a2147fd821906612/fhd.mp4   
36  https://cdn-st.rutubelist.ru/media/ab/e5/a77cf9a94a78b621c7c754942b29/fhd.mp4   
37  https://cdn-st.rutubelist.ru/media/f7/41/65b82be24517a0e37298975a3b07/fhd.mp4   
38  https://cdn-st.rutubelist.ru/media/89/0f/84a8a71b4d249365ae19fbd1fcff/fhd.mp4   
39  https://cdn-st.rutubelist.ru/media/a6/8c/3c8921b34ba8ae905ec143136b22/fhd.mp4   
40  https://cdn-st.rutubelist.ru/media/f0/6f/d0cce3454eb89b70209b67a4d970/fhd.mp4   
41  https://cdn-st.rutubelist.ru/media/e5/30/d9df4fb940939da69c0b351fa170/fhd.mp4   
42  https://cdn-st.rutubelist.ru/media/f3/1e/ff08193b402296920e2a1995bd16/fhd.mp4   
43  https://cdn-st.rutubelist.ru/media/70/4b/1d7a08354550904b96695fc51c01/fhd.mp4   
44  https://cdn-st.rutubelist.ru/media/45/c4/88b971f7414e94d7c7f3265fc160/fhd.mp4   
45  https://cdn-st.rutubelist.ru/media/94/3d/ea40da9645acacf006df16b9bb75/fhd.mp4   
46  https://cdn-st.rutubelist.ru/media/f1/1b/e51475c34340b472c9757326e0a0/fhd.mp4 

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400000 entries, 0 to 399999
Data columns (total 2 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   link         400000 non-null  object
 1   description  345081 non-null  object
dtypes: object(2)
memory usage: 6.1+ MB


In [5]:
rows_count = dataset['link'].count()
na_rows_count = dataset[dataset['description'].isna()]['link'].count();

print("Общее количество записей {}".format(rows_count))
print("Количество записей без тегов {}".format(na_rows_count))
print("Процент записей без тегов {:.0%}".format(na_rows_count/rows_count))

Общее количество записей 400000
Количество записей без тегов 54919
Процент записей без тегов 14%


После первой попытки обобщенного анализа имеем:
- Датасет состоит из 400 000 записей
- Одна запись содержит линку на видео, и тэги, актуальные для нее
- Имеем, что ~55 000 записей (14%) не имеют тегов, в принципе

Попробуем понять, какой объем данных мы имеем. Это важно для того, чтобы выработать стратегию обучения. Мы знаем, что сервис организатора представляет собой платформу для просмотра коротких видео. Следовательно, можно предположить, что размер контента +- одинаковый. То есть, для примерного подсчета размера выборки будем придерживаться следующей логики:
- С помощью get запроса по рандомно взятому link получим видео и взглянем на размер ответа в header'ах
- Полученный размер возьмем за среднее. Не будем заниматься тем, что реально посчитаем средний размер по выборке или же полностью просканим 400к записей для того, чтобы получить ответ. Выглядит излишним и шумным. На конечный результат, судя по вводной относительно логики работы платформы, данная точность повлияет с небольшим размахом, который в реальности не имеет смысла.
- Умножим средний размер на количество записей, зафиксируем полученный результат
- Порассуждаем над стратегией работы с данными

In [6]:
sizes = []
for i in range(0, 10):
    test_link = dataset['link'][i]
    response = requests.get(test_link)
    size = int(response.headers['Content-Length']) / 1024 / 1024
    sizes.append(size)
    
meanSize = np.array(sizes).mean()

print(f"Средний размер файла {meanSize:.2f} мб")

meanVolume = meanSize * 400000 / 1024 / 1024


print(f"Средний размер выборки {meanVolume:.2f} тб")

Средний размер файла 6.09 мб
Средний размер выборки 2.32 тб


Итак, имеем, что прогнозируемый размер выборки - 2.32тб. Объем достаточно внушительный для того, чтобы работать с данными дефолтным образом и локально. Ощущение, что обработку видео придется осуществлять батчами (пачками). Также необходимо учесть, что само видео непосредственно получаем GET'ом с хоста, который предоставили организаторы. И здесь нужно отметить, что:
- в каких-то местах процесса обогащения можем падать по rate limiter'у
- хост может быть перегружен, наверняка команды соперников так же будут к нему обращаться
- возможны задержки в отдаче в пиковые моменты, надо подумать над увеличением таймаутов 
- подумать о ретраях и вообще обработке exception'ов от хоста

И прежде чем перейти к обработке взглянем на то, а какое количество видео имеет идентичные хэштеги

In [31]:
unique_values = dataset['description'].unique()
unique_values_percent = unique_values.shape[0]/ dataset['description'].shape[0]
print('Количество уникальных описаний в виде хештегов к видео:',unique_values.shape[0])
print('Процент уникальных описаний в виде хештегов к видео:',  f"{unique_values_percent:.0%}")

Количество уникальных описаний в виде хештегов к видео: 91075
Процент уникальных описаний в виде хештегов к видео: 23%


Данная статистика и сессия ответов на вопросы с экспертами по треку навела на мысль о том, что в выборке могут быть дубли по контенту. Очевидно, что вряд ли корректным равенством является видео 1 = видео 2, если равны хэштеги, но, на всякий случай, было решено, что:
- будем смотреть на 1 фрейм видео
- считать хеш от него, и проверять, есть ли в кэше результаты работы моделей по видео. 
- В случае, если хеши совпали, будем пропускать обработку данного видео и обогощать его из кэша

Теперь перейдем к описанию стратегии по обработке

## Стратегия обработки видео

Начнем от простого к сложному, поэтому распишу максимально абстрактно. Предпологаем, что будем действовать следующим образом:
- Брать n видео из выборки
- Смотрим, есть ли описание по видео в кэше, если нет, то следуем дальше по списку вниз
- Полученные видео прогоняем через whisper, в надежде получить транскрибацию с видео
- Сэмплируем каждое видео на m картинок
- Строим архитектуру под Image Captioning. Предварительно будем брать архитектуры, которые обучены на Flickr или COCO 2017ого года, и файнтюнить их, в идеале брать обученные
- Перед тем, как начать основной флоу, возьмем первый кадр с видео и глянем на его хеш
- В случае, если данный хеш обратывался ранее - возьмем фичи для видео из кэша
- Прогоняем полученные m картинок, получаем результат, переводим его и объединяем все в одну строку
- Полученные описания (текст с whisper, текст с описания изображения) складываем в БД. Предварительно будем использовать Elastic. Также открытый вопрос


Дальнейшие действия будем осуществлять на Kaggle, так как у каждого разработчика MAC OS и для использования GPU нужны дополнительные команды. Тестово начнем удаленно, посмотрим на скорость работы заданного алгоритма. Если все будет плохо и обработка будет занимать длительное время, перепишем код под локальное железо и продолжим на MAC OS.
Преждем чем продемонстрировать пример работы скрипта по обогащению, зафиксируем, из чего состоит сбор фичей компонентно:

1. **Image Captioning** - собираем описание контента с видео и переводим его на русский язык
2. **Automatic Speech Recognition** - собираем речь с видео в текст, если таковая присутсвует
3. **Text to Embedding** - полученное переведенное описание и речь переводим в векторный вид с помощью OpenSource модели от SberDevice. Также используем исходные теги, видно, что признак, в принципе надежный (всего у 14% записей теги отсутствуют вовсе) и этот признак так же можно учесть в поиске

Далее смотреть тетрадку **preprocess_description_with_examples**, в которой можете отражен пример работы пайплайна и описаны проблемы, с которыми сталкивались, а так же каким образом их решали. В тетрадке много текста, но просьба прочитать =)

После того, как работа процесса в отдельной тетрадке рассмотрена, перейдем к описанию полученных результатов

## Полученные результаты и выводы

Итак, взглянем полученную выборку, которую использования для построения индекса

In [33]:
from pandas import option_context

with option_context('display.max_colwidth', 400, "display.max_rows", 100):
    general_data = pd.read_csv('../general_data.csv')
    display(general_data.head(100))

link  \
0   https://cdn-st.rutubelist.ru/media/b0/e9/ef285e0241139fc611318ed33071/fhd.mp4   
1   https://cdn-st.rutubelist.ru/media/39/6c/b31bc6864bef9d8a96814f1822ca/fhd.mp4   
2   https://cdn-st.rutubelist.ru/media/e9/e0/b47a9df14a5e97942715e5e705c0/fhd.mp4   
3   https://cdn-st.rutubelist.ru/media/87/43/b11df3f344d0af773aac81e410ee/fhd.mp4   
4   https://cdn-st.rutubelist.ru/media/d1/e7/642dc2194fcdb69664f832d5f2dd/fhd.mp4   
5   https://cdn-st.rutubelist.ru/media/e2/97/f9164f8a41479f961d64842154a7/fhd.mp4   
6   https://cdn-st.rutubelist.ru/media/0f/48/8a1ff7324073947a31e80f71d001/fhd.mp4   
7   https://cdn-st.rutubelist.ru/media/19/9c/98de3b4a4d9e83cb5dda828f899e/fhd.mp4   
8   https://cdn-st.rutubelist.ru/media/bf/ee/85fb6b80491db79e8baebe5c9d80/fhd.mp4   
9   https://cdn-st.rutubelist.ru/media/fc/73/d86600d04e519178a55774dfe65e/fhd.mp4   
10  https://cdn-st.rutubelist.ru/media/17/ae/a36f6f824fa1a67496d299c8ac6e/fhd.mp4   
11  https://cdn-st.rutubelist.ru/media/4b/54/0a36f96d4030a0f6ad1b6d7f4902/fhd.mp4   
12  https://cdn-st.rutubelist.ru/media/8c/58/96608b7a4834883ea2dd1c884c63/fhd.mp4   
13  https://cdn-st.rutubelist.ru/media/2f/4f/6b969c8c4a2aafcfca057b2a99a2/fhd.mp4   
14  https://cdn-st.rutubelist.ru/media/9f/ca/d5d0dedf44418e1a2b7d5cebe0e8/fhd.mp4   
15  https://cdn-st.rutubelist.ru/media/a4/d2/8e9737fd445faf0a20151f8c39f5/fhd.mp4   
16  https://cdn-st.rutubelist.ru/media/a3/9f/2352de2748b3868df583d51a402b/fhd.mp4   
17  https://cdn-st.rutubelist.ru/media/ac/08/3604604b4b8b89e34db1e0e91863/fhd.mp4   
18  https://cdn-st.rutubelist.ru/media/5a/35/65c80c3047d08634922cb22f1fee/fhd.mp4   
19  https://cdn-st.rutubelist.ru/media/03/eb/47329b1943349255453b6cec625d/fhd.mp4   
20  https://cdn-st.rutubelist.ru/media/28/b0/88ac63a546cb8174020a95470591/fhd.mp4   
21  https://cdn-st.rutubelist.ru/media/bf/6a/f040f0dd4afc90b8eb12c8d76571/fhd.mp4   
22  https://cdn-st.rutubelist.ru/media/73/48/47a9e25b4b36bc6254d29f42787a/fhd.mp4   
23  https://cdn-st.rutubelist.ru/media/fe/86/ab9d0a504c2a892c88dc02b79b56/fhd.mp4   
24  https://cdn-st.rutubelist.ru/media/86/0f/4a6b06514445bbe41cb62a5f279d/fhd.mp4   
25  https://cdn-st.rutubelist.ru/media/1e/39/f6e45a2142b095d96b286744441a/fhd.mp4   
26  https://cdn-st.rutubelist.ru/media/0d/97/531d4356476a9deb64239aed758a/fhd.mp4   
27  https://cdn-st.rutubelist.ru/media/85/3c/4ade487d46cfb36205c43e4dca40/fhd.mp4   
28  https://cdn-st.rutubelist.ru/media/50/0a/a56eb2824b93a57c3f4ad7e02ddd/fhd.mp4   
29  https://cdn-st.rutubelist.ru/media/a7/ff/3d901b774fc8b5728833abe9485d/fhd.mp4   
30  https://cdn-st.rutubelist.ru/media/b3/e2/07e47c1e4af8a1ea0a932b5cdae3/fhd.mp4   
31  https://cdn-st.rutubelist.ru/media/39/8a/ce3d8db142e38d99c708f406cca0/fhd.mp4   
32  https://cdn-st.rutubelist.ru/media/c5/74/92d9fe2a4fd99509c72ebf46e7d7/fhd.mp4   
33  https://cdn-st.rutubelist.ru/media/95/65/d45056f64cd48e919324c2ade667/fhd.mp4   
34  https://cdn-st.rutubelist.ru/media/19/b5/aa6a2f7b46ad81e3c78e947d472f/fhd.mp4   
35  https://cdn-st.rutubelist.ru/media/03/b6/7aab45314f66a2147fd821906612/fhd.mp4   
36  https://cdn-st.rutubelist.ru/media/ab/e5/a77cf9a94a78b621c7c754942b29/fhd.mp4   
37  https://cdn-st.rutubelist.ru/media/f7/41/65b82be24517a0e37298975a3b07/fhd.mp4   
38  https://cdn-st.rutubelist.ru/media/89/0f/84a8a71b4d249365ae19fbd1fcff/fhd.mp4   
39  https://cdn-st.rutubelist.ru/media/a6/8c/3c8921b34ba8ae905ec143136b22/fhd.mp4   
40  https://cdn-st.rutubelist.ru/media/f0/6f/d0cce3454eb89b70209b67a4d970/fhd.mp4   
41  https://cdn-st.rutubelist.ru/media/e5/30/d9df4fb940939da69c0b351fa170/fhd.mp4   
42  https://cdn-st.rutubelist.ru/media/f3/1e/ff08193b402296920e2a1995bd16/fhd.mp4   
43  https://cdn-st.rutubelist.ru/media/70/4b/1d7a08354550904b96695fc51c01/fhd.mp4   
44  https://cdn-st.rutubelist.ru/media/45/c4/88b971f7414e94d7c7f3265fc160/fhd.mp4   
45  https://cdn-st.rutubelist.ru/media/94/3d/ea40da9645acacf006df16b9bb75/fhd.mp4   
46  https://cdn-st.rutubelist.ru/media/f1/1b/e51475c34340b472c9757326e0a0/fhd.mp4 

Визуально пробежав по небольшой выборке видно, что далеко не у каждой записи имеется разговорная часть в видео. 
Взглянем на это статистически

In [36]:
general_data_length = general_data.shape[0]
general_data_notna_length = general_data[general_data['text'].notna()].shape[0]

In [38]:
print("Общее количество записей {}".format(general_data_length))
print("Количество записей без разговора на видео {}".format(general_data_notna_length))
print("Процент записей без разговора на видео {:.0%}".format(general_data_notna_length/general_data_length))

Общее количество записей 50
Количество записей без разговора на видео 38
Процент записей без разговора на видео 76%


Видно, что большая часть видео из исходной выборки не содержит разговора на видео. Это говорит о том, что вырабатывая стратегию поиска, на данный атрибут нужно ориентироваться в последнюю очередь и приоритетом брать эмбединги от description_ru

Что же, выработка фичей описана, данную работу считаем завершенной. В дальнейшем перейдем к подробному описанию поиска